rfc/0078-zone-aware-replica-reads.md

```diff
@@ -0,0 +1,346 @@
+In this case, all reads will be done only from the local availability zone. It
+is the cheapest solution, although it reduces chances of getting data from
+replicas (if the server groups are unbalanced).
+
+![Selected Server Group](figures/0078-case-2-local-only.svg)
+
+## Selected Server Group or All Available
+
+This strategy allows using all available replicas, but only in case when the
+local server group is empty. The selected server group might be empty in case
+the groups are not balanced properly on the cluster, or some nodes has been
+failed over. The previous strategy in this case would return
+`102 DocumentIrretrievable` error and refuse touching replicas from non-local
+group.
+
+Note that in this case we do not do second round trip after trying nodes from
+selected server group, but only use current topology to detect the Empty-Group
+case.
+
+![Selected Server Group or All Available](figures/0078-case-3-selected-server-group-or-all-available.svg)
+
+# Changes
+
+## Updates in Configuration Parser
+
+Each node in `nodesExt` array of cluster configuration will have new property
+`serverGroup`, which contains name of the group as it seen in Admin UI. For
+example:
+
+```jsonc
+{
+    // ...
+    "vBucketServerMap": {
+        "vBucketMap": [
+            [ 0, 1 ]
+            // ... other vBuckets
+        ]
+    },
+    // ...
+    "nodesExt": [
+        {
+            // ...
+            "hostname": "172.17.0.2",
+            "serverGroup": "group_1"
+        },
+        {
+            // ...
+            "hostname": "172.17.0.3",
+            "serverGroup": "group_2"
+        },
+        {
+            // ...
+            "hostname": "172.17.0.4",
+            "serverGroup": "group_2"
+        },
+        {
+            // ...
+            "hostname": "172.17.0.5",
+            "serverGroup": "group_1"
+        }
+    ]
+    // ...
+}
+```
+
+Having this information, the SDK will be able to filter list of the node
+indexes in `vBucketMap` to get local server group members.
+
+For example, lets say the configured SDK to use `"group_1"` as its local server
+group. Then if some key is mapped to vBucket `0`, the SDK can filter vector if
+indexes `[0, 1]` to only `[0, -1]` (using `-1` here for illustration), because
+server with index `1` belongs to `"group_2"`. In the result, the SDK only need
+to send `GET` request to retrieve active copy of the document.
+
+## Selecting Server Group for Connection
+
+To allow user to select preferred server groups, new setter/property for
+`ClusterOptions` must be introduced:
+
+```
+class ClusterOptions {
+    // ...
+    ClusterOptions preferredServerGroup(String serverGroupName);
+    // ...
+}
+```
+
+Where `serverGroupName` matches name of the group as it seen in Admin UI of the
+server and in cluster configuration JSON.
+
+## Selecting Read Preference for Operations
+
+New enumeration should be defined for expressing different strategies for the
+read replica APIs:
+
+```
+enum ReadPreference {
```

Comment:
Slight nit:  SELECTED_SERVER_GROUP at least is a strict command rather than a preference.  Should this be ReadRequirement?

I don't feel strongly on this though.

Reply:
My though process was like this:

1. the connection options have it `preferredServerGroup` as the setting is global, and I want to avoid impression that it applies to *all* operations
2. then the enumeration describes different type of this preference, so technically later we can expand it with more sophisticated strategies like the one I described in the initial draft were we do cascaded requests: first try local group, and then if it gives us nothing -- do the regular request automatically without using try/catch block

Reply:
Ok, fair enough.  It's a good point that it aligns with the setting.